### DISCUSSION TITLE
Homer7-prom stack - Network effectiveness ratio BY number

### DISCUSSION THREAD

**Discussion by 3klips** [sipcapture/homer#651] [Q&A] [answered]
Hello guy,

I'm running in my lab the Homer 7 stack with prometheus, timescale-postgress.
With Grafana connected to Prometheus, I can have a great graph of the NER (Network effectiveness ratio), you can filter this graph by node.

My challenge is to have the same report filtered by calling or called number filter.

How would you resolve this as the calling/called number is not part of the Prometheus database.
Using homer API?

Thanks!
Enric

**Comment by adubovikov** [comment, marked as answer]
Hi [@3klips](https://github.com/3klips),

Good question. Short answer: **Prometheus NER by node works; NER by calling/called number does not come from the prom stack out of the box.** Use the **Homer database/API** for number-level reporting.

### Why Prometheus cannot filter by number today

In **Homer 7 prom** (`heplify-server` → Prometheus → Grafana), SIP KPIs come from counters like `heplify_method_response` with labels such as:

- `target_name` (from `PromTargetIP` / `PromTargetName`)
- `node_id`
- `method`
- `response`
- `direction`

There is **no `from_user` / `to_user` / E.164 label** by design. Adding every phone number as a Prometheus label would explode cardinality and destabilize heplify-server (this has bitten people before).

So your Grafana NER panel filtered by **node** is the intended prom use case. Filtering by **number** is a different data path.

### Practical options for NER by number (Homer 7)

**Option A — Grafana + Timescale/PostgreSQL (recommended for dashboards)**

Keep Prometheus for node-level KPIs. Add a **PostgreSQL/Timescale** datasource pointing at the same DB Homer uses (`hep_proto_1_call`).

Example pattern (adjust column names to your schema/version):

```sql
SELECT
  time_bucket('1 hour', create_date) AS time,
  100.0 * COUNT(*) FILTER (WHERE method = 'INVITE' AND reply = '200')
        / NULLIF(COUNT(*) FILTER (WHERE method = 'INVITE'), 0) AS ner_pct
FROM hep_proto_1_call
WHERE create_date BETWEEN $__timeFrom() AND $__timeTo()
  AND (
    from_uri ILIKE '%' || '${number:raw}' || '%'
    OR to_uri   ILIKE '%' || '${number:raw}' || '%'
  )
GROUP BY 1
ORDER BY 1;
```

Use a Grafana template variable `${number}` for the caller/callee filter.

This gives you the same kind of time-series graph, but sourced from stored SIP CDR rows, not Prometheus counters.

**Option B — Homer API / search**

Use Homer search/stat APIs to pull INVITE transactions for a number/time window, then compute NER in your app or script.

Useful for ad-hoc troubleshooting, less ideal for a live dashboard unless you build a small aggregator.

**Option C — Pre-aggregate a small allowlist**

If you only need NER for a **fixed set** of numbers (e.g. 10 test DIDs), a sidecar job can compute NER periodically and expose custom Prometheus metrics with a `number` label.

Do **not** do this for all subscribers in production.

### How this relates to `PromTargetIP` / `PromTargetName`

Those settings map **capture node / trunk IP** → friendly `target_name`. They help filter by **network element**, not by subscriber number.

Same idea as your current node filter, one layer lower in the network.

### Homer 11 note

Homer 11 exposes similar aggregate SIP Prometheus metrics (`homer_sip_method_response_total`, also keyed by `target_name`). Per-number KPIs still belong in **SIP search / SQL over stored calls**, not in high-cardinality prom labels.

### Summary

| Report | Best source |
|--------|-------------|
| NER by node/trunk | Prometheus (`heplify_method_response`, `target_name`) |
| NER by calling/called number | Timescale/PostgreSQL on `hep_proto_1_call` or Homer API |
| Mixed dashboard | Grafana with **two datasources**: Prometheus + Postgres |

If you share your Grafana NER query and Homer/heplify-server version, we can adapt the SQL to your exact column names.
